### ISSUE TITLE
publishedActionsEvaluator should be able to show types using Template Haskell

### ISSUE THREAD

**Issue by Hi-Angel** [yi-editor/yi#865]
Feature request: help message *(e.g. typing :help in vim mode)* could be more useful if after names it would include according types.

**Comment by jhance** [comment]
@Hi-Angel Do you have Hint enabled?

**Comment by jhance** [comment]
Marking as difficult because `publishedActions` is not known at compiletime.

**Comment by jhance** [comment]
For this to work, we need to directly put the types into the `publishedActions` config variable. We need a `PublishedAction' datatype and need a function `publishAction :: String -> PublishedAction`.

**Comment by Hi-Angel** [comment]
@jhance how do I enable hint? [I found here](https://hackage.haskell.org/package/yi) that by `-f hint` flag, but it gives me `yi: Panic: frontend not found`.

**Comment by jhance** [comment]
That should work... maybe try `-f hint -f vty`

**Comment by Hi-Angel** [comment]
@jhance nope, the same output.

**Comment by jhance** [comment]
Thats quite puzzling

**Comment by Hi-Angel** [comment]
@jhance yep; in fact I'm on my new system, and I installed Haskell just yesterday to compile yi. So I can even list actions out of my memory *(okay, and bash history)*:

* sudo pacman -S ghc cabal
* cabal update
* git clone git://github.com/yi-editor/yi.git
* cd yi
* cabal install

    *(here I got a little error)*
* cabal install alex
* cabal install
* cp example-configs/yi-vim.hs ~/.config/yi/

Btw, I'm wondering, how would I execute functions inside Yi? E.g. how would I execute `printMsg "hello"` in editor?

**Comment by jhance** [comment]
@Hi-Angel You could run like `:yi printMsg "hello"`. However without hint support you may be limited in what functions you can run :(

I'm going to try switching my yi installation to the `master` branch and see what happens.

**Comment by Hi-Angel** [comment]
I'm not sure if that relevant, but I got here `ghc-8.0.1` to which I didn't even manage to find changelog, only a status page.

**Comment by jhance** [comment]
We should have ghc8 support iirc, although the stack config I've been using is locked to lts.

**Comment by jhance** [comment]
Hmm, I can reproduce this issue on master. I do `:help pointB` and see `pointB :: BufferM Point`. I used stack instead of cabal. This is my stack file:

```
resolver: lts-6.8

# User packages to be built.
# Various formats can be used as shown in the example below.
# 
# packages:
# - some-directory
# - https://example.com/foo/bar/baz-0.0.2.tar.gz
# - location:
#    git: https://github.com/commercialhaskell/stack.git
#    commit: e7b331f14bcffb8367cd58fbfc8b40ec7642100a
# - location: https://github.com/commercialhaskell/stack/commit/e7b331f14bcffb8367cd58fbfc8b40ec7642100a
#   extra-dep: true
#  subdirs:
#  - auto-update
#  - wai
# 
# A package marked 'extra-dep: true' will only be built if demanded by a
# non-dependency (i.e. a user package), and its test suites and benchmarks
# will not be run. This is useful for tweaking upstream packages.
packages:
- yi-language
- yi
extra-deps: []
flags: {}
```

I build against `yi-language` master but that should not be an issue.

**Comment by Hi-Angel** [comment]
Okay, I'm not at home ATM, I'll try when I got there to build it again with stack.

**Comment by Hi-Angel** [comment]
@jhance well, I'm in office managed to chroot into an Archlinux :Ь I didn't build with stack yet, am I understand right: I should copy the code you posted into a `yi.yaml`, and run `stack build` ?

**Comment by noughtmare** [comment]
@Hi-Angel No, to build with stack you should probably run `stack init` and then `stack install --flag yi:hint --flag yi:vty` (the `--flag yi:hint` and `--flag yi:vty` are enabled by default and should not be necessary). 

You can also try to run `stack install yi --flag yi:hint --flag yi:vty` which installs Yi from [stackage](https://www.stackage.org/).

**Comment by Hi-Angel** [comment]
Thank you!

Okay, so I compiled it with `stack install` *(i.e. without `--flag yi:hint --flag yi:vty`)*, and I see the same problem. I'll try in a minute to recompile it again with the flags *(I'm not sure, do I need to run something to clean binary files, or a simple restart would be enough?)*.

**Comment by noughtmare** [comment]
You can run `stack clean` to be sure that everything is cleaned, but I don't think it's necessary.

**Comment by Hi-Angel** [comment]
Anyway, I ran `stack clean`, then `stack install --flag yi:hint --flag yi:vty`, and I still getting the error

	~/.local/bin/yi  -f hint -f vty
	yi: Panic: frontend not found

**Comment by noughtmare** [comment]
You shouldn't run `yi` with `-f hint -f vty`. Yi should be run with just `yi [file]`. Check `yi --help` for all options. (The -f option is for selecting the frontend that is why it couldn't find the frontend).

**Comment by Hi-Angel** [comment]
Well, turns out, `stack install` have its own problems: when I start yi without options, I see errors like `Could not find module `Yi'`. I didn't see those errors when compiled with cabal, I guess it needs some paths to be set.

**Comment by Hi-Angel** [comment]
Aha, I figured it out! It's `stack exec yi`!

**Comment by Hi-Angel** [comment]
Okay, so… Now, when I'm pressing `:` *(in vim keybindings)*, and typing any `help` or `yi`, it says `Unknown command`. The cabal build, at least, has `help` executed.

Also, [per comments](https://github.com/yi-editor/yi/issues/865#issuecomment-234835222), commands with explicit `hint` are triggering errors:

	$ stack exec yi -- -f vty -f hint
	Run from outside a project, using implicit global project config
	Using resolver: lts-5.13 from implicit global project's config file: /home/constantine/.stack/global-project/stack.yaml
	yi: Panic: frontend not found
	$ stack exec yi -- -f hint
	Run from outside a project, using implicit global project config
	Using resolver: lts-5.13 from implicit global project's config file: /home/constantine/.stack/global-project/stack.yaml
	yi: Panic: frontend not found

**Comment by Hi-Angel** [comment]
Oh, I'm sorry, never mind, the last time I ran commands were in the wrong tab, it was in the host system, not the chrooted one…

**Comment by Hi-Angel** [comment]
Anyway, the problem with *frontend panic* still persist, and trying to execute `yi printMsg "hello"` results in:

	error: WontCompile [GhcError {errMsg = "<interactive>:2:1:\n    No instance for (Show x0) arising from a use of `makeAction'\n    The type variable `x0' is ambiguous\n    Note: there are several potential instances:\n      instance Show a => Show (Maybe a) -- Defined in

Then it goes beyond the screen.

**Comment by noughtmare** [comment]
The `-f hint` and `-f vty` are compile time flags. They only need to be added when you are building Yi with cabal. The reason that Yi is complaining about missing the frontend is because the `-f frontend` is a run time argument of Yi. Yi thinks you want to use the `hint` frontend (which doesn't exist).

**Comment by noughtmare** [comment]
While playing around with `:yi` I found out that there are still a lot of problems . I don't think it is ready for use.

**Comment by noughtmare** [comment]
Also the example-configs are outdated and need to be updated.

**Comment by noughtmare** [comment]
`:yi printMsg "hello"` should be `:yi EditorA $ printMsg "hello"` (see #872) but hint has issues (see #870) so it wont work yet. The I split the rest of the issues in this thread into their own issues.

**Comment by jhance** [comment]
The core issue (of not being able to show types in some cases) still remains and isn't one of the separate issues you created.

**Comment by noughtmare** [comment]
If hint is disabled all published actions have type `Action` right?

**Comment by noughtmare** [comment]
We can show types with `typeOf` (from `Data.Typeable`).

**Comment by noughtmare** [comment]
Damn, that close button is in the same place as the cancel button...

**Comment by jhance** [comment]
I feel like if you close and reopen in 5 seconds github should just not say you did either. It happens alot
